langserver/workspace_refs.go

```diff
@@ -77,65 +104,90 @@ func (h *LangHandler) handleWorkspaceReference(ctx context.Context, conn JSONRPC
-// package and returns the results.
-func (h *LangHandler) externalRefsFromPkg(ctx context.Context, bctx *build.Context, conn JSONRPC2Conn, pkg string, rootPath string, results *refResultSorter) (err error) {
-	span, ctx := opentracing.StartSpanFromContext(ctx, "externalRefsFromPkg")
+func (h *LangHandler) externalRefsTypecheck(ctx context.Context, bctx *build.Context, conn JSONRPC2Conn, fset *token.FileSet, pkgs []string) (prog *loader.Program, err error) {
+	span, ctx := opentracing.StartSpanFromContext(ctx, "externalRefsTypecheck")
 	defer func() {
 		if err != nil {
 			ext.Error.Set(span, true)
 			span.SetTag("err", err.Error())
 		}
 		span.Finish()
 	}()
-	span.SetTag("pkg", pkg)
 
-	// Import the package.
-	buildPkg, err := bctx.Import(pkg, rootPath, 0)
-	if err != nil {
-		if !(strings.Contains(err.Error(), "no buildable Go source files") || strings.Contains(err.Error(), "found packages") || strings.HasPrefix(pkg, "github.com/golang/go/test/")) {
-			return fmt.Errorf("skipping possible package %s: %s\n", pkg, err)
-		}
-		return nil
+	// Configure the loader.
+	var typeErrs []error
+	conf := loader.Config{
+		Fset: fset,
+		TypeChecker: types.Config{
+			DisableUnusedImportCheck: true,
+			FakeImportC:              true,
+			Error: func(err error) {
+				typeErrs = append(typeErrs, err)
+			},
+		},
+		Build:       bctx,
+		AllowErrors: true,
+		ParserMode:  parser.AllErrors | parser.ParseComments, // prevent parser from bailing out
+		FindPackage: func(bctx *build.Context, importPath, fromDir string, mode build.ImportMode) (*build.Package, error) {
+			// When importing a package, ignore any
+			// MultipleGoErrors. This occurs, e.g., when you have a
+			// main.go with "// +build ignore" that imports the
+			// non-main package in the same dir.
+			bpkg, err := bctx.Import(importPath, fromDir, mode)
+			if err != nil && !isMultiplePackageError(err) {
+				return bpkg, err
+			}
+			return bpkg, nil
+		},
 	}
-
-	if strings.HasPrefix(buildPkg.Dir, "vendor/") || strings.Contains(buildPkg.Dir, "/vendor/") {
-		// Per the workspace/reference docs:
-		//
-		// 	- Excluding any `URI` which is located within the workspace.
-		// 	- Excluding any `Location` which is located in vendored code (e.g.
-		// 	  `vendor/...` for Go, `node_modules/...` for JS, .tgz NPM packages, or
-		// 	  .jar files for Java).
-		//
-		// This means that we do not need to consider vendor directories at all
-		// since we do not emit references that are inside the same workspace
-		// (vendor always is) and do not emit references to things that are
-		// vendored. Thus, we can skip typechecking the entire vendor directory
-		// which saves us a a lot of work.
-		return nil
+	for _, path := range pkgs {
+		conf.Import(path)
 	}
 
-	pkgInWorkspace := func(path string) bool {
-		return PathHasPrefix(path, h.init.RootImportPath)
+	// Load and typecheck the packages.
+	prog, err = conf.Load()
```

Comment:
@keegancsmith the magic is actually here. This only gets called once, for all packages (see `conf.Import(path)` above) rather than once for each package as it was before.

Basically, keeping only one `loader.Config` for all your request-scoped work is a good idea :)